### PR TITLE
#240 Fix RefResolver and make it compatible with draft-04

### PR DESCRIPTION
**Pull request by jojo1981**
The public API of the RefResolver will be changed. Suggesting to take this PR in a new Major version of this library.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,4 @@ coverage
 .settings
 composer.lock
 docs-api
+.idea
```
Comment (bighappyface) [inline .gitignore]: You should add this to a global ignore: http://stackoverflow.com/questions/7335420/global-git-ignore

**Comment by bighappyface** [comment]
@araines what do you think of this alternative approach to resolving the ref issue (pun intended)?

File: .gitignore (hunk 1 above)
Comment (jojo1981) [inline .gitignore]: I agree. I will update the code and revert this change!

**Comment by jojo1981** [comment]
@bighappyface last weekend a good friend of mine also a great programmer was at my home. We've both have studied again on my solution and have improved the code. I have squashed the commits and did a new push. Can you please re-check this code?

**Comment by bighappyface** [comment]
@jojo1981 sorry I didn't chime in on this sooner.

There is a set of ref test provided by the JSON Schema test suite that are currently skipped because of the missing functionality.

You can enable those tests by omitting this line:

https://github.com/justinrainbow/json-schema/blob/61a95a151aac86eaa9bbd31311d3d12a78c971b8/tests/JsonSchema/Tests/Drafts/Draft4Test.php#L21

I pulled this PR down and ran the test suite with those ref tests enabled and ran into a lot of failures. I would expect those tests to pass first and foremost to improve our draft 4 coverage.

**Comment by araines** [comment]
@bighappyface I'm happy to review this once its working with the tests, also I'm not precious about which solution gets used moving forward - so long as the library is improving, thats all that matters!

**Comment by jojo1981** [comment]
@bighappyface Good point I will try to fix the code so the draft 4 tests will also work.

**Comment by jojo1981** [comment]
@bighappyface You're right, those tests also needed to be executed and of course pass successfully. I will fix it and push it again.

**Comment by jojo1981** [comment]
@all Updated the code and enabled test.

**Comment by jojo1981** [comment]
@bighappyface I have ommitted more than only this line: https://github.com/justinrainbow/json-schema/blob/61a95a151aac86eaa9bbd31311d3d12a78c971b8/tests/JsonSchema/Tests/Drafts/Draft4Test.php#L21

**Comment by jojo1981** [comment]
Rebased from `master` branch.

**Comment by jojo1981** [comment]
@araines I agree! Same for me I don't care which solution is moved forward as long as the library improves.

**Comment by araines** [comment]
Picking it up and taking a look :)

**Comment by araines** [comment]
@jojo1981 - Interesting how we independently have approached the problem in a very similar manner - I guess that is probably a positive thing!  However, I think this solution has the edge - it feels a bit cleaner and neater than the one I did in #210 .  The only downside to this one is that we are changing the documented way of using the library, which means it will break for more people (rather than those who had custom-extended it).

However, having looked over this and weighed it up, I think this represents the better solution, even though it will cause slightly more pain for people using this library elsewhere.  @bighappyface I'd personally recommend taking this PR over #210

**Comment by bighappyface** [comment]
@mirfilip would you mind taking a look at this?

**Comment by jojo1981** [comment]
@araines I agree it's not backwards compatible and can only be used in the next major (I think this is a good step for this library). Thanks for your positive remarks! For now I will leave this PR and do no forced pushed anymore. I'm waiting for review comment. So people let's start reviewing!

File: src/JsonSchema/RefResolver.php [hunk 2]
```diff
@@ -9,269 +9,175 @@
...
+    private function getRefSchema(JsonPointer $jsonPointer, $schema)
     {
-        if (! isset($schema->$propertyName) || ! is_object($schema->$propertyName)) {
-            return;
+        /** @var object|array $refSchema */
+        $refSchema = $schema;
+        foreach ($jsonPointer->getPropertyPaths() as $path) {
+            if (is_object($refSchema) && property_exists($refSchema, $path)) {
+                $refSchema = $refSchema->{$path};
+            } elseif (is_array($refSchema) && array_key_exists($path, $refSchema)) {
```
Comment (araines) [inline src/JsonSchema/RefResolver.php]: How does this work with getting the last array element (a tilde)?  Spec here: https://tools.ietf.org/html/rfc6901#section-4

I couldn't see a test for it, and I'm trying to wrap my head around how this bit works.

**Comment by araines** [comment]
@jojo1981 given the Interfaces are now defined and are nice and simple, does that mean the other public methods of UriResolver and UriRetriever should potentially be made protected/private - or perhaps the interfaces should be extended to include these if they are necessary?  If my memory serves me correctly, there is a fair bit of duplication between the UriResolver and UriRetriever which we can perhaps slim out whilst you are making public interface breaking changes anyway?

File: src/JsonSchema/RefResolver.php (hunk 2 above)
Comment (jojo1981) [inline src/JsonSchema/RefResolver.php]: The JsonSchema\Entity\JsonPointer should implement the logic as describe in: https://tools.ietf.org/html/rfc6901#section-4

**Comment by jojo1981** [comment]
@araines I agree, but suggest doing that in a separate issue. This PR is unintentional getting bigger than I at first expected. Because this is a library it's always a good idea to use simple and clean interfaces. Only have some public methods who has a clean purpose. Maybe after the complete refactoring we should use annotations on public methods like: `@api` when it's a stable public method and `@deprecated` for methods which will removed in the next major.

File: src/JsonSchema/RefResolver.php (hunk 2 above)
Comment (araines) [inline src/JsonSchema/RefResolver.php]: I see that is the intention, but as I understand it the `JsonPointer` only splits the path elements, meaning you'd end up with a list of the various parts.  That is fine, and is used here, but it looks like at that point the `$refSchema` would be an array and the code highlighted would be executed, looking for the path of `~` in the reference schema - I don't understand how that would ensure we end up with the last element of the array (I may well be missing something).

Either way, I would suggest adding a test for this particular test case to ensure it works and in future the functionality is not broken when further work is done.

**Comment by araines** [comment]
@jojo1981 - fair enough, i'm sure it can be handled as a separate matter.

File: src/JsonSchema/RefResolver.php (hunk 2 above)
Comment (jojo1981) [inline src/JsonSchema/RefResolver.php]: I agree I have to re-check this part. I have re-runned all currently existing PHP Units tests and see this part is not fully covered (lines 108, 109 and 110 are not hit by any test).

Also I think It's better to throw an exception when `$refSchema` can not be fully resolved. Now the current `$schema` will be returned (see line 110).

screen shot 2016-04-06 at 21 12 23

Also I will try to provide a test which coverage line 108.

Comment (jojo1981) [inline src/JsonSchema/RefResolver.php]: After some more investigation is see line 108 is dead code. So I will fix this.

**Comment by jojo1981** [comment]
Updated the code but with an extra commit. When this review is finished and can be merged I will squash all the commits and force push again. For now while active in the review process I think is handier to have separate commits.

**Comment by jojo1981** [comment]
@mirfilip would you mind also taking a look at this, just as @bighappyface asked?

**Comment by jojo1981** [comment]
@bighappyface did you re-check this PR?

**Comment by bighappyface** [comment]
@jojo1981 sorry, I have been monitoring the thread but haven't jumped into the full review quite yet. I'll jump on it this weekend for sure.

**Comment by jojo1981** [comment]
@bighappyface No problemo!

File: src/JsonSchema/RefResolver.php [hunk 3]
```diff
@@ -9,269 +9,177 @@
...
      */
-    public function setUriRetriever(UriRetriever $retriever)
-    {
-        $this->uriRetriever = $retriever;
-
-        return $this;
-    }
-
-    protected function resolveRefSegment($data, $pathParts)
+    private function getReservedKeysWhichAreInFactSubSchemas()
```
Comment (bighappyface) [inline src/JsonSchema/RefResolver.php]: Interesting name here. Very matter-of-fact. I don't have a problem with it, just commenting on the choice of words/phrasing.

File: tests/JsonSchema/Tests/Constraints/BaseTestCase.php [hunk 4]
```diff
@@ -10,24 +10,39 @@
...
-        $refResolver = new RefResolver(new UriRetriever);
-        $refResolver->resolve($schema);
+        $value = json_decode($input);
 
         $validator = new Validator($checkMode);
+        $validator->check($value, $schema);
 
-        $validator->check(json_decode($input), $schema);
+        if ($validator->isValid()) {
+            $a = 2;
```
Comment (bighappyface) [inline tests/JsonSchema/Tests/Constraints/BaseTestCase.php]: Not sure what is going on here.

File: tests/JsonSchema/Tests/Drafts/Draft3Test.php [hunk 5]
```diff
@@ -12,11 +25,12 @@ protected function getFilePaths()
...
     }
 
+    /**
+     * {@inheritdoc}
+     */
     protected function getSkippedTests()
     {
         return array(
-            'ref.json',
-            'refRemote.json',
```
Comment (bighappyface) [inline tests/JsonSchema/Tests/Drafts/Draft3Test.php]: BOOM! :boom:

File: tests/JsonSchema/Tests/Drafts/Draft4Test.php [hunk 6]
```diff
@@ -12,18 +25,15 @@ protected function getFilePaths()
...
+     * {@inheritdoc}
+     */
     protected function getSkippedTests()
     {
         return array(
-            // Not Yet Implemented
-            'definitions.json',
-            // Partially Implemented
-            'ref.json',
-            'refRemote.json',
```
Comment (bighappyface) [inline tests/JsonSchema/Tests/Drafts/Draft4Test.php]: BOOM! :boom: 

Really stoked about this expansion of features for this package.

**Comment by bighappyface** [comment]
@jojo1981 @araines this looks pretty solid to me. I'll admit that I didn't get too deep into the logic of these changes, but I tend to trust developers, particularly when they write tests to support their claims :)

As far as I am concerned this is good to go and merits a major bump. Do you concur?

File: tests/JsonSchema/Tests/Constraints/BaseTestCase.php (hunk 4 above)
Comment (jojo1981) [inline tests/JsonSchema/Tests/Constraints/BaseTestCase.php]: Ugly... just some stupid debug code. I will remove this.

File: src/JsonSchema/RefResolver.php (hunk 3 above)
Comment (jojo1981) [inline src/JsonSchema/RefResolver.php]: :-) Yeah I know. A better name is welcome. So if you have a suggestion?

**Comment by jojo1981** [comment]
Processed the remarks and added an extra commit.

**Comment by jojo1981** [comment]
@bighappyface When we good to go just let me know so I can squash the commits.

**Comment by bighappyface** [comment]
@jojo1981 go ahead and squash the commits. I'm slammed today but I will work on the merge and bump this evening.

**Comment by jojo1981** [comment]
@bighappyface I've squashed the commits as you can see.

**Comment by bighappyface** [comment]
+1

**Comment by bighappyface** [comment]
@jojo1981 @araines we are at 2.0.0:

https://github.com/justinrainbow/json-schema/releases/tag/2.0.0

**Comment by jojo1981** [comment]
@bighappyface Thx!

**Pull request merged**